src/main/game/CommandLineUI.java

```diff
@@ -47,16 +47,28 @@ public String getValidMove(Board board) {
         return move;
     }
 
-    public void displayBoard(List<List<String>> rows) {
+    public void displayBoard(List<Mark> grid) {
         StringBuilder sb = new StringBuilder();
         String formattedRows = null;
-        for (int index = 0; index <= rows.size() - 1; index++) {
-            List<String> row = rows.get(index);
-            sb.append(row);
-            sb.append("\n");
+
+        for (int i = 0; i < grid.size(); i++) {
+            if (grid.get(i) == Mark.EMPTY) {
+                sb.append(" ");
+                sb.append(i + 1);
+            } else {
+                sb.append(grid.get(i));
+            }
+
+            if (i == 2) {
+                sb.append("\n");
+            } else if (i == 5) {
+                sb.append("\n");
+            } else if (i == 8) {
+                sb.append("\n");
+            }
```

Comment:
Good job for getting this to work. Try and see if there's a pattern here that can be generalised so that you avoid using the hardcoded line boundaries (2, 5, 8).